### PR TITLE
SONAR-7480 optimize SQL requests of QualityProfile search ws

### PR DESCRIPTION
**Pull request by sns-seb**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java [hunk 1]
```diff
@@ -220,6 +247,32 @@ public int compare(QProfile o1, QProfile o2) {
     }
   }
 
+  private enum LanguageToKey implements Function<Language, String> {
+    INSTANCE;
+
+    @Override
+    @Nonnull
+    public String apply(@Nonnull Language input) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java [hunk 2]
```diff
@@ -81,105 +88,125 @@ SearchData load(SearchWsRequest request) {
...
-      dbClient.closeSession(dbSession);
+    addAll(qualityProfiles, findDefaultProfiles(dbSession, languageKeys));
+    return difference(languageKeys, qualityProfiles.keySet());
+  }
+
+  private static <T> Set<T> difference(Set<T> languageKeys, Set<T> set2) {
+    return Sets.newHashSet(Sets.difference(languageKeys, set2));
+  }
+
+  private void addAllFromDto(Map<String, QProfile> qualityProfiles, Collection<QualityProfileDto> list) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java]: MAJOR Make "addAllFromDto" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java [hunk 3]
```diff
@@ -81,105 +88,125 @@ SearchData load(SearchWsRequest request) {
...
-   * If no profile found, try to find the quality profile set on the project (if provided)
-   * If still no profile found, try to find the default profile of the language
-   * <p/>
-   * Never return null because a default profile should always be set on each language
-   */
-  private QProfile getProfile(DbSession dbSession, String languageKey, @Nullable String projectKey, @Nullable String profileName) {
-    QualityProfileDto profileDto = profileName != null ? profileFactory.getByNameAndLanguage(dbSession, profileName, languageKey) : null;
-    if (profileDto == null && projectKey != null) {
-      profileDto = profileFactory.getByProjectAndLanguage(dbSession, projectKey, languageKey);
+  private void addAll(Map<String, QProfile> qualityProfiles, Collection<QProfile> list) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java]: MAJOR Make "addAll" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java [hunk 4]
```diff
@@ -81,105 +88,125 @@ SearchData load(SearchWsRequest request) {
...
-    return profileDtoToQProfile(profileDto);
+  private Set<String> getLanguageKeys() {
+    return from(Arrays.asList(languages.all())).transform(LanguageToKey.INSTANCE).toSet();
   }
 
-  private QProfile getDefaultProfile(DbSession dbSession, String languageKey, @Nullable String profileName) {
-    QualityProfileDto profile = profileName != null ? profileFactory.getByNameAndLanguage(dbSession, profileName, languageKey) : null;
-    if (profile == null) {
-      profile = profileFactory.getDefault(dbSession, languageKey);
+  private List<QProfile> findDefaultProfiles(Set<String> languageKeys) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java]: MAJOR Private method 'findDefaultProfiles' is never used. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java [hunk 5]
```diff
@@ -220,6 +247,32 @@ public int compare(QProfile o1, QProfile o2) {
...
+      return input.getKey();
+    }
+  }
+
+  private enum QualityProfileDtoToQProfile implements Function<QualityProfileDto, QProfile> {
+    INSTANCE;
+
+    @Override
+    @Nonnull
+    public QProfile apply(@Nonnull QualityProfileDto input) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/SearchDataLoader.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

File: sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java [hunk 6]
```diff
@@ -273,6 +297,16 @@ public QualityProfileDto selectByNameAndLanguage(String name, String language, D
     return mapper(session).selectByNameAndLanguage(name, language);
   }
 
+  public List<QualityProfileDto> selectByNameAndLanguages(final String name, Collection<String> languageKeys, final DbSession session) {
+    return DatabaseUtils.executeLargeInputs(languageKeys, new Function<List<String>, List<QualityProfileDto>>() {
+      @Override
+      @Nonnull
+      public List<QualityProfileDto> apply(@Nonnull List<String> input) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

File: sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java [hunk 7]
```diff
@@ -198,6 +212,16 @@ public QualityProfileDto selectByProjectAndLanguage(DbSession session, String pr
     return mapper(session).selectByProjectAndLanguage(projectKey, language);
   }
 
+  public List<QualityProfileDto> selectByProjectAndLanguages(final DbSession session, final String projectKey, Collection<String> languageKeys) {
+    return DatabaseUtils.executeLargeInputs(languageKeys, new Function<List<String>, List<QualityProfileDto>>() {
+      @Override
+      @Nonnull
+      public List<QualityProfileDto> apply(@Nonnull List<String> input) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

File: sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java [hunk 8]
```diff
@@ -169,7 +172,18 @@ public void delete(int id) {
   }
 
   @CheckForNull
-  public QualityProfileDto selectDefaultProfile(DbSession session, String language) {
+  public List<QualityProfileDto> selectDefaultProfiles(final DbSession session, Collection<String> languageKeys) {
+    return DatabaseUtils.executeLargeInputs(languageKeys, new Function<List<String>, List<QualityProfileDto>>() {
+      @Override
+      @Nonnull
+      public List<QualityProfileDto> apply(@Nonnull List<String> input) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/qualityprofile/QualityProfileDao.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* MAJOR 8 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.

**Comment by teryk** [comment]
LGTM

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 5 major

Watch the comments in this conversation to review them.

**Pull request merged**